### DISCUSSION TITLE
daemon.json official documentation in rootless mode ?

### DISCUSSION THREAD

**Discussion by gautaz** [moby/moby#47058] [Q&A] [answered]
Hello,

I have been trying to find some official documentation about the `daemon.json` file in **rootless** mode.
The only reference I have found so far is here: https://docs.docker.com/engine/security/rootless/#docker-pull-errors

Has anyone found more extensive documentation about this file in rootless mode ?

In particular, I am trying to add a runtime in rootless mode by following this documentation:
https://docs.docker.com/engine/reference/commandline/dockerd/#configure-runtimes-using-daemonjson

As it does not seem to work, I was wondering if all `daemon.json` options are available in rootless mode.

**Comment by thaJeztah** [comment, marked as answer]
> As it does not seem to work, I was wondering if all daemon.json options are available in rootless mode.

Do you have more details what doesn't work?

I _do_ notice that we could improve some of the documentation about _where_ the config file is expected to be stored. The [daemon configuration file](https://docs.docker.com/engine/reference/commandline/dockerd/#daemon-configuration-file) section mentions default locations for Linux and Windows, but I _think_ for rootless we use a different location (`~/.config/docker/daemon.json`, but I _think_ `XDG_CONFIG_DIR` is also taken into account). And for Docker Desktop, we (currently) store a copy in `~/.docker/daemon.json` (which is synced with the actual configuration inside the VM).

Perhaps we should at least add a table somewhere that documents the (default) location for various known situations (Linux, Windows, rootless (with / without `XDG_CONFIG_DIR` set), Docker Desktop)

/cc @dvdksn

**Reply by dvdksn** [reply]
On it

**Reply by gautaz** [reply]
Hello @thaJeztah, this morning, while answering to another issue from another user (https://discourse.nixos.org/t/docker-rootless-with-nvidia-support/37069/4), I realized that the way docker rootless is deployed on my Linux distribution (NixOS) uses a "non-standard" path for the `daemon.json` file.

So modifying `~/.config/docker/daemon.json` had no effect, hence my questions on this matter.
Can you confirm that adding docker runtimes to the rootless `daemon.json` file is supposed to work as intended?
Now that I have the right path, I should be able to fix the issue I have with the nvidia runtime in rootless mode.

**Reply by AkihiroSuda** [reply]
> Can you confirm that adding docker runtimes to the rootless daemon.json file is supposed to work as intended?

Yes, nvidia container toolkit works with Rootless Docker, at least on Ubuntu.

For Ubuntu 22.04, I just had to do:
- Create `~/.config/docker/daemon.json` and run `systemctl --user restart docker`
```json
{
    "runtimes": {
        "nvidia": {
            "args": [],
            "path": "nvidia-container-runtime"
        }
    }
}
```

- Modify `/etc/nvidia-container-runtime/config.toml`
```diff
--- /etc/nvidia-container-runtime/config.toml.BAK	2024-01-16 07:02:05.606573439 +0000
+++ /etc/nvidia-container-runtime/config.toml	2024-01-16 07:02:11.114549694 +0000
@@ -10,7 +10,7 @@
 #ldcache = "/etc/ld.so.cache"
 ldconfig = "@/sbin/ldconfig.real"
 load-kmods = true
-#no-cgroups = false
+no-cgroups = true
 #path = "/usr/bin/nvidia-container-cli"
 #root = "/run/nvidia/driver"
 #user = "root:video"
```

**Reply by AkihiroSuda** [reply]
Merge request for https://docs.nvidia.com/datacenter/cloud-native/container-toolkit/latest/install-guide.html

- https://gitlab.com/nvidia/cloud-native/cnt-docs/-/merge_requests/373
